Mathport/Syntax/Translate/Basic.lean

```diff
@@ -168,37 +206,109 @@ def renameModule (n : Name) : M Name := do Rename.renameModule (← read).pcfg n
+def mkIdentN (n : Name) : M Syntax := do mkIdentR (← renameNamespace n)
+def mkIdentF (n : Name) : M Syntax := do mkIdentR (← renameField n)
+def mkIdentO (n : Name) : M Syntax := do mkIdentR (← renameOption n)
 
 def Parser.ParserM.run' (p : ParserM α) (args : Array (Spanned VMCall)) : M α := do
   match p.run ⟨(← read).commands, args⟩ with
   | Except.ok a => pure a
   | Except.error e => throw! "unsupported: {e}"
 
+def mkCommentString (comment : Comment) : String :=
+  if comment.text.contains '\n' then s!"/-{comment.text}-/\n" else s!"--{comment.text}\n"
+
+def addLeadingComment' (comment : Comment) (info : SourceInfo) : SourceInfo :=
+  let commentText := mkCommentString comment
+  match info with
+    | SourceInfo.none =>
+      SourceInfo.original commentText.toSubstring (positionToStringPos comment.start) "".toSubstring (positionToStringPos comment.end)
+    | SourceInfo.synthetic a b =>
+      SourceInfo.original commentText.toSubstring a "".toSubstring b
+    | SourceInfo.original leading a trailing b =>
+      SourceInfo.original (commentText ++ leading.toString).toSubstring a trailing b
+
+partial def addLeadingComment (comment : Comment) (stx : Syntax) : Option Syntax :=
+  if let Syntax.node i k args := stx then Id.run do
+    for j in [0:args.size] do
+      if let some a' := addLeadingComment comment args[j] then
+        return Syntax.node i k (args.set! j a')
+    pure none
+  else
+    stx.setInfo (addLeadingComment' comment stx.getInfo)
+
+def addTrailingComment' (comment : Comment) (info : SourceInfo) : SourceInfo :=
+  let commentText := mkCommentString comment
+  match info with
+    | SourceInfo.none =>
+      SourceInfo.original "".toSubstring (positionToStringPos comment.start) commentText.toSubstring (positionToStringPos comment.end)
+    | SourceInfo.synthetic a b =>
+      SourceInfo.original "".toSubstring a commentText.toSubstring b
+    | SourceInfo.original leading a trailing b =>
+      SourceInfo.original leading a (trailing.toString ++ commentText).toSubstring b
+
+partial def addTrailingComment (comment : Comment) (stx : Syntax) : Option Syntax :=
+  if let Syntax.node i k args := stx then Id.run do
+    for j in [0:args.size] do
+      let j := args.size - j - 1
+      if let some a' := addTrailingComment comment args[j] then
+        return Syntax.node i k (args.set! j a')
+    pure none
+  else
+    stx.setInfo (addTrailingComment' comment stx.getInfo)
+
+def nextCommentIf (p : Comment → Bool) : M (Option Comment) := do
+  let firstComment :: remainingComments := (← get).remainingComments | return none
+  unless p firstComment do return none
+  modify ({ · with remainingComments })
+  return firstComment
+
+partial def insertComments (stx : Syntax) : M Syntax := do
+  if let some headPos := stx.getInfo.getPos? then
+    if let some comment ← nextCommentIf (positionToStringPos ·.«end» ≤ headPos) then
+      let stx ← insertComments stx
+      let stx := (addLeadingComment comment stx).getD stx
```

Comment:
This drops the comment if it can't find a place to put it in the syntax. How about pushing the comment back in the queue in this case?

Reply:
Good idea!